### DISCUSSION TITLE
Efficientdet_lite3_512 as edgetpu_model.tflite

### DISCUSSION THREAD

**Discussion by Minglarn** [blakeblackshear/frigate#2336] [Ask A Question]
I saw 3 new tflite tpu files as corals homepage...
![image](https://user-images.githubusercontent.com/58695742/143785740-365ee89c-f134-4e25-a9a6-ccd1fc2bad68.png)
![image](https://user-images.githubusercontent.com/58695742/143785809-1263c21d-a10a-4d0a-8e3d-8a6d4b587a45.png)

Does the lite3 have higher accuracy at (mAP2) 39.4% or do I read this completely wrong ?

**Comment by blakeblackshear** [comment]
Yes, but it is 10x slower

**Reply by ozett** [reply]
does "slower" matter on 6ms per frame on a CORAL? or while having two CORLAs?
do i break anything? or can i use the model straight away in my existing config? i am tempted to play with it and change the model setting...(if i manage to find the right place in the config-file 😄 )

can i use this? i try to download it manually and see if i find the right place to put it...
(https://github.com/google-coral/test_data/blob/master/efficientdet_lite3_512_ptq_edgetpu.tflite)

**Reply by ozett** [reply]
is there a way to check if frigate is really using this new model?
i had no log-errors, but inference-times now as high as with CPU
anything else looks normal... but i am a bit unsure, if frigate framework makes use of this model.. 
a visual clue or a special debug-section would be fine for this..
![image](https://user-images.githubusercontent.com/5937811/143841035-209aaaa6-c12a-4b02-82fb-b4c8edb1ed52.png)

**Reply by blakeblackshear** [reply]
It's working. That's what I mean by 10x slower.

**Reply by Minglarn** [reply]
> does "slower" matter on 6ms per frame on a CORAL? or while having two CORLAs? do i break anything? or can i use the model straight away in my existing config? i am tempted to play with it and change the model setting...(if i manage to find the right place in the config-file 😄 )
> 
> can i use this? i try to download it manually and see if i find the right place to put it... (https://github.com/google-coral/test_data/blob/master/efficientdet_lite3_512_ptq_edgetpu.tflite)

5 frames per sec * 100ms ... = half a second ... Fair enough for me... :)
It works fine for me and i'll let it run for a week just to be sure...

**Reply by Minglarn** [reply]
> Yes, but it is 10x slower

Thanks...

**Reply by ozett** [reply]
![image](https://user-images.githubusercontent.com/5937811/143890768-9fefde4b-473e-43cc-8365-29b03bdb7d95.png)
i will stay with it for a while and also see how it goes...

**Reply by ozett** [reply]
> 5 frames per sec * 100ms ... = half a second ... Fair enough for me... :)

with a 5 FPS for detection-cam, does frigate have 1000ms for all 5 frames - 
and the coral can use roughly a duration of max. 200ms per Frame for detections? 

i have two CORALS... can i load one model in one coral? and another model to the other - 
whith the same timeframe assumptions?
or can they do a round-robin to get an even larger (double?) timeframes for slower models?

questions over questions, but experimenting can also be fun, 
so i updated to a more 😄  time-consuming model an see how this goes 
(more prescision is less false-positives, i guess...)

![image](https://user-images.githubusercontent.com/5937811/143913707-45f154e1-86ce-4441-96c7-e72e8875ed1a.png)
==
![image](https://user-images.githubusercontent.com/5937811/143914300-786cbf72-6db6-4f77-bddb-052d65daae82.png)

**Reply by Minglarn** [reply]
You cheater... :) 
So you got a m2 card? Looking for a m2 card my self. I've only got 2 USB Corals... Witch model do you have?
![image](https://user-images.githubusercontent.com/58695742/143930295-12e71e79-7aaf-4227-aa6f-e88e624febb8.png)

**Reply by ozett** [reply]
> Witch model do you have?

i bought some m2 corals before they went over night out of stock since the last weeks..
i also bought a dual coral, but still needs a pci-adapter to get both working. pre-ordered one from https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter

**Comment by fcasarsa** [comment]
in some ( unfortunately often ) occasions i have some false positives. Could be possible to do a dual - stage inference?

eg use a lighter model for "real time" and then a double check with a more cpu-intensive model?

Thanks.

**Comment by Minglarn** [comment]
It sure is more accurate...
New 
![image](https://user-images.githubusercontent.com/58695742/143928818-c2569484-e0fe-45a5-abaf-01e5fc84bf9d.png)
vs 
Old
![image](https://user-images.githubusercontent.com/58695742/143928866-7fe58e59-7205-4df9-9de9-320660310408.png)

**Reply by ozett** [reply]
> It sure is more accurate...

more accurate on the same object, means only higher percentage of beeing an object of the learned object-set.
but couldnt also mean "more accurate" => beeing more  accurat on more detecting of a greater variety of (different) objects from prior unknown image informations? somehow invisible whats "Not" detected from a lack of accuracy.. such would be hard to measure?

**Comment by ozett** [comment]
i had the [model ](https://coral.ai/models/object-detection/)```efficientdet_lite3x_640_ptq_edgetpu.tflite``` run overnight.
my impression is, that there had been less detections.
with the [default model](https://github.com/blakeblackshear/frigate/blob/v0.9.4/docker/Dockerfile.base#L43) ```ssdlite_mobiledet_coco_qat_postprocess_edgetpu``` i got ~70 detections, with the ```lite3x``` only 7...
i switched back to the default and think how to measure a model comparison.. 🤔 

at daylight the ```lite3x```  seems to work better than the default. 
dont know how to do an illustrating shoutout between modeln in real usage scenario.. 🤔 

edit: maybe evalute models this way? -> https://voxel51.com/docs/fiftyone/tutorials/evaluate_detections.html

**Reply by Minglarn** [reply]
I've experienced the opposite.... Got far more recordings at night time compared with the fast tflite file. 
It gets dark at 15:00 here in Sweden so there are more activity outside when it's dark. 

Maybe you should tweak your settings...

**Reply by ozett** [reply]
> Maybe you should tweak your settings...

for comparison i did not want to change settings,
but i can optimize settings for ```lite3x```model at nighttime. but clueless what will help... some hints?

**Comment by ozett** [comment]
What about [this new ones](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md)?
funny model name...anybody wants to try?
![image](https://user-images.githubusercontent.com/5937811/144377738-a3708b5c-03cd-4b80-9309-e675c9f7c454.png)

**Reply by blakeblackshear** [reply]
I tested them. Didn't see a meaningful difference in speed or accuracy in my test set.

**Reply by ozett** [reply]
great, that calms me down... 😄 
but i still suspect uncanny 👻 recognition rates and new objects 👻 from other models, especially from the large tensorflow models. 😉  What is the best way to compare this in a test? can i adopt your test setup? Is it complex? or easy, so i can try to build a comparison setup?

**Comment by ozett** [comment]
yolo3 and yolo5 on coral-tpu...-> https://github.com/google-coral/edgetpu/issues/405#issuecomment-984848877
worth a try?

**Comment by ozett** [comment]
20 objects, 
person, cat, dogs ... really suitable for surveillance ..

what about coloring persons after detection with first model?
this could be run after first fast detection on CORAL in a 2nd stage on CPU.

would be great fun to see person in color - in frigate 👍 

![image](https://user-images.githubusercontent.com/5937811/144564120-a2e7e888-51c6-46ec-a75f-c5c56a00a8f0.png)

**Comment by Minglarn** [comment]
Found this page... Apperantly a HUb with a lot of models, trained for varius objects... 
https://tfhub.dev/

Based on COCO 2017
https://tfhub.dev/tensorflow/collections/object_detection/1

Going to test some of them...

**Comment by ozett** [comment]
yeah great, just one minute ago i found out, that you could run tensoflow in a docker to tryout the models.
times ago it was a real pain to install all dependencies. but that looks better now

https://www.tensorflow.org/install/docker
and than try deeplap segmentation on CPU
https://github.com/tensorflow/models/tree/master/research/deeplab

next days i must find a way to get the research-models into this docker (as i am more a n00b than a coder)

**Reply by gururise** [reply]
I think most of those models are trained on images taken in the day.  Would be great to have a model trained on both day/night images and IR images.

**Reply by ozett** [reply]
there are cctv-image set to train on for optimized model. 
https://github.com/blakeblackshear/frigate/discussions/1043#discussioncomment-1522414
think the key is a simple way to plug (custom trained- or new) models into frigate.
i have hopes that this could be done with a tensorflow-docker container..

edit: dockerize new models -> https://www.tensorflow.org/tfx/serving/docker?hl=en

**Reply by ozett** [reply]
running a small test for a segmentation-AI on a snapshot of a saved frigate-event.
looks somehow cool, specially for night images. could be a start for imaging ideas... 😄 


<kbd> ![image](https://user-images.githubusercontent.com/5937811/145079223-1fc645ca-0a41-4523-84b2-4449422ca8dc.png)
https://github.com/google-research/deeplab2/blob/main/g3doc/setup/installation.md

<kbd> ![image](https://user-images.githubusercontent.com/5937811/145081530-c08f500b-7896-4884-a63e-b93610cf6f72.png) </kbd> 

start to tweak the colormap 😄 

<kbd>![image](https://user-images.githubusercontent.com/5937811/145083023-0b762469-56a5-485d-9ea9-a85d9dcb7c0f.png) </kbd> 

the segmentation-ai fails on some full snapshots, where frigate detects perfectly..maybe 2nd level segmentation for color-mapping must be run only on the that area-part, where frigate has detected an object... one idea to test...

**Comment by ozett** [comment]
deeplab running for testing on frigate detections event-ids...
looks some kindof cool to me...

![image](https://user-images.githubusercontent.com/5937811/145795451-da17b026-9e57-4cc7-9907-da546ec0060a.png)
